File: src/workflows/workflowDependencyEngine.ts
/**
 * Converted from Python: core/workflows/workflow_dependency_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function buildWorkflowDependencies(plan: any): any {
  var ordering: any[] = [];
  var prerequisites: any[] = [];
  var runtime_deps: any[] = [];
  var chains: any[] = [];
  var steps: any = [...py.iter(py.get(plan, "steps", []))];
  var index: any;
  var step: any;
  for ([index, step] of py.enumerate(py.slice(steps, null, 10000))) {
    var depends_on: any = py.toStr(py.get(step, "depends_on", ""));
    if (py.truthy(depends_on)) {
      py.listAppend(ordering, {"step": py.toStr(py.get(step, "id", "")), "depends_on": depends_on});
      py.listAppend(prerequisites, {"step": py.toStr(py.get(step, "id", "")), "requires": depends_on});
    }
    if ((index > 0)) {
      var prev_runtime: any = py.toStr(py.get(py.at(steps, py.sub(index, 1)), "runtime", ""));
      var curr_runtime: any = py.toStr(py.get(step, "runtime", ""));
      if (!py.eq(prev_runtime, curr_runtime)) {
        py.listAppend(runtime_deps, {"from": prev_runtime, "to": curr_runtime});
      }
    }
    py.listAppend(chains, {"step": py.toStr(py.get(step, "id", "")), "action": py.toStr(py.get(step, "action", ""))});
  }
  return {"execution_ordering": ordering, "semantic_prerequisites": prerequisites, "runtime_dependencies": runtime_deps, "extraction_chains": chains, "bounded": true};
}

File: lib/src/workflows/workflow_dependency_engine.dart
// Port of core/workflows/workflow_dependency_engine.py

Map<String, dynamic> buildWorkflowDependencies(
  Map<String, dynamic> plan,
) {
  final List<Map<String, dynamic>> ordering = <Map<String, dynamic>>[];
  final List<Map<String, dynamic>> prerequisites = <Map<String, dynamic>>[];
  final List<Map<String, dynamic>> runtimeDeps = <Map<String, dynamic>>[];
  final List<Map<String, dynamic>> chains = <Map<String, dynamic>>[];

  final List<dynamic> steps = plan['steps'] is List
      ? List<dynamic>.from(plan['steps'] as List<dynamic>)
      : <dynamic>[];
  final List<dynamic> bounded = steps.take(10000).toList();

  for (int index = 0; index < bounded.length; index++) {
    final dynamic stepRaw = bounded[index];
    final Map<String, dynamic> step = stepRaw is Map
        ? Map<String, dynamic>.from(stepRaw)
        : <String, dynamic>{};
    final String dependsOn = '${step['depends_on'] ?? ''}';
    if (dependsOn.isNotEmpty) {
      ordering.add(<String, dynamic>{
        'step': '${step['id'] ?? ''}',
        'depends_on': dependsOn,
      });
      prerequisites.add(<String, dynamic>{
        'step': '${step['id'] ?? ''}',
        'requires': dependsOn,
      });
    }

    if (index > 0) {
      final dynamic prevRaw = bounded[index - 1];
      final Map<String, dynamic> prev = prevRaw is Map
          ? Map<String, dynamic>.from(prevRaw)
          : <String, dynamic>{};
      final String prevRuntime = '${prev['runtime'] ?? ''}';
      final String currRuntime = '${step['runtime'] ?? ''}';
      if (prevRuntime != currRuntime) {
        runtimeDeps.add(<String, dynamic>{
          'from': prevRuntime,
          'to': currRuntime,
        });
      }
    }

    chains.add(<String, dynamic>{
      'step': '${step['id'] ?? ''}',
      'action': '${step['action'] ?? ''}',
    });
  }

  return <String, dynamic>{
    'execution_ordering': ordering,
    'semantic_prerequisites': prerequisites,
    'runtime_dependencies': runtimeDeps,
    'extraction_chains': chains,
    'bounded': true,
  };
}
